common/predictive-text/workerFactory.ts

```diff
@@ -0,0 +1,5 @@
+namespace com.keyman.text.prediction {
+  export interface WorkerFactory {
```

Comment:
I'm not sure I understand the benefits of using the factory pattern over:

 - a callback that returns something that implements the `Worker` interface
 - using dependency injection to pass something implementing the `Worker` interface to the LMLayer?

Could you please elaborate why this is necessary?

Reply:
> I'm not sure I understand the benefits of using the factory pattern over:
> 
>     * a callback that returns something that implements the `Worker` interface

Uh... so... a factory pattern?  I don't see the difference here, but maybe that's due to my somewhat heavy Java-teaching background.  Factory = class that exists to implement exactly this callback.

Then again, I never did take a formal class on design principles, so maybe I'm using the "factory" name wrong.  Most "patterns" I've learned have either been from intuition or analyzing ones I've seen in the wild.

----

Rather than go on more rabbit trails after that...

While the DOM-based `Worker` is extremely simple to set up, the synthetic `VirtualizedWorker` seen in the next PR is fairly complex.  That said, upon another look... yeah, we could just have the derived classes pass in a fresh constructed instance instead of a callback...  but I'd then make a `Worker` subclass that accomplishes the same goal noted [here](https://github.com/keymanapp/keyman/pull/2986#discussion_r408902720), which could then mirror the `VirtualizedWorker` class.  I very much like having the `Worker` setup relegated to its own class; it helps keep things nice and modular.

That gives me a few ideas for refining the next PR in the chain; I should probably keep `VIrtualizedWorker` separate from the lm-layer specific setup code in there, defining that as part of its own layer.  But that's a different PR, anyway.

Reply:
I think part of the reasoning is that the two Worker types will require a different setup.  The DOM-based worker is initialized with a `Blob`, but we can't use that type without DOM access.  So, the 'factory' was to reflect that difference.

Reply:
When does the setup of the Worker need to happen?

Reply:
?  It never does.  The `Worker` is always fully instantiated before the `LMLayer` constructor completes.

Note that `LMLayerBase`'s constructor expects to interact with a live, initialized `Worker` - it seeks to immediately send it a `config` message, if I recall correctly.

Reply:
In that case, I would shy away from introducing factory classes with a type hierarchy, when a simple callback can suffice. A factory class type-hierarchy increases code bloat, which I know @mcdurdin wishes to prevent. In TypeScript/JavaScript, many design patterns from the Gang of Four's original patterns for C++/SmallTalk/Java that consist of an interface containing a single method can be converted to a [function type](https://www.typescriptlang.org/docs/handbook/interfaces.html#function-types), with minimal overhead in the compiled output.

But since the Worker  (whether a WebWorker or a "virtualized" worker) can be instantiated before instantiating the `LMLayer` I don't even see the purpose of doing that.

Reply:
Our current design for initializing the `LMLayer` allows consumers to simply initialize it without parameters, right?  The `LMLayer` then performs a "default" initialization, setting up a default `Worker`.

Of course, that "default `Worker`" is currently DOM-dependent.  We'd need a different default when the DOM is not available; in fact, the `LMLayer` cannot compile without DOM references when that default is included.

So, this provides the way to make two different child classes; the only real difference is that they provide different "defaults" specific to their target platform / compilation strategy.

Short form:  if `LMLayer` code is responsible for setting up a `Worker` if none is provided, we need a mechanism to ensure the correct "default" is provided.

Reply:
I have a different opinion: this PR changes the signature of the `LMLayer` constructor. The reason the `LMLayer` has been responsible for instantiating the worker, is because, if the worker is not provided as an optional argument, then the LMLayer's constructor will create it for you, as an API convenience. **The LMLayer is not intrinsically responsible for constructing the Worker**. As such, since the signature of the `LMLayer` constructor is going to change anyway, the more lightweight change is to change the signature to _require_ an instantiated worker, rather than create a new type hierarchy to fulfill this purpose.

Reply:
OK... if we're fine with that design change, things can become a _lot_ simpler.  I'm pretty wary of making changes to external API when possible; that's probably influence from the rest of the team.

Reply:
Is this an external API though? External API changes need to be thoroughly documented but with internal API changes it's up to the user to update when they depend on those.

Reply:
By "external API" do you mean one that is intended to be used by end-users? Because this one isn't really intended to be used by your everyday Keyman developer user; it is not even required when creating custom models. It is just required when instantiating the LMLayer. Both the changes from when the original PR was uploaded, and the changes after I requested changes change the API of instantiating the LMLayer — so this will be used by KeymanWeb and perhaps anybody using the LMLayer standalone (i.e., me 😂); but I think the way Joshua wrote the API now is much cleaner.

Reply:
Yeah, an "internal API" is one that we formalize for inter-module connections (whatever a 'module' may be). It may be documented on our wiki for our purposes. But it isn't presented as a documented, stable interface for people outside the team to use. An example of an internal API is the [downloads.keyman.com bundles API](https://github.com/keymanapp/keyman/wiki/downloads.keyman.com-bundles-API).

Sometimes we may decide to publish an internal API to an external API, as we discover good use cases (or someone asks us to). All our external APIs are documented on help.keyman.com.

At least that's the theory.